libcob/java.c

```diff
@@ -0,0 +1,183 @@
+*/
+
+#include <jni.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+// For caching which is an optimization but can we internally implement it as well?
+typedef struct {
+    jclass cls;
+    jmethodID mid;
+} MethodCache;
+
+MethodCache* cache = NULL;
+int cacheSize = 0;
+
+static JavaVM *jvm = NULL;    
+static JNIEnv* env = NULL; // pointer to native method interface
+
+static void 
+cob_add_to_cache(jclass cls, jmethodID mid) {
+    cacheSize++;
+    cache = (MethodCache*)realloc(cache, sizeof(MethodCache) * cacheSize);
+    cache[cacheSize - 1].cls = cls;
+    cache[cacheSize - 1].mid = mid;
+}
+
+jmethodID 
+cob_get_from_cache(jclass cls, const char* methodName, const char* methodSig) {
+    for (int i = 0; i < cacheSize; i++) {
+        if (cache[i].cls == cls) {
+            return cache[i].mid;
+        }
+    }
+    return;
+}
+
+static JNIEnv* 
+cob_create_vm() {
+    JavaVMInitArgs args; // JDK/JRE 6 VM initialization arguments
+    JavaVMOption* options = (JavaVMOption*)cob_malloc(sizeof(JavaVMOption) * 1);
+    args.version = JNI_VERSION_1_6;
+    args.nOptions = 1; //inline
+    options[0].optionString = "-Djava.class.path=/usr/lib/java";
+    args.options = &options;
+    args.ignoreUnrecognized = 1;
+    int rv;
+    rv = JNI_CreateJavaVM(jvm, (void**)&env, &args); // loading and initializing a Java VM, returning as JNI interface
+    if (rv < 0 || !env)
+        return;
+    else
+        return rv;
+    return env;
+}
+
+static void 
+cob_handle_error(JavaVM* jvm, char* methodSig) {
+    if (methodSig != NULL) {
+        free(methodSig);
+    }
+    if (jvm != NULL) {
+        (*jvm)->DestroyJavaVM(jvm);
+    }
+}
+
+static char* 
+cob_gen_method_sig(const char** paramType, int paramCount, const char** returnType) {
+    // (param_list) + return_type
+    int length = 2 + strlen(returnType);
+    for(int i = 0; i < paramCount; i++) {
+        length += strlen(paramType[i]);
+    }
+    
+    // internal malloc
+    char* sig = (char*)cob_malloc(length + 1);
+    sig[0] = '(';
+    int pos = 1;
+    for(int i = 0; i < paramCount; i++) {
+        strcpy(sig + pos, paramType[i]);
+        pos += strlen(paramType[i]);
+    }
+    sig[pos++] = ')';
+    strcpy(sig + pos, returnType);
+    return sig;
+}
+
+static void
+cob_lookup_static_method(JNIEnv* env, JavaVM* jvm, const char *className, const char *methodName, const char *methodSig, const char *returnType, const char** paramTypes, int paramCount) {
+    jclass cls = (*env)->FindClass(env, className);
+    if (cls == NULL) {
+        printf("Class not found\n");
+        (*jvm)->DestroyJavaVM(jvm);
+        return;
+    }
+    
+    jmethodID mid = get_from_cache(cls, methodName, methodSig);
+    if (mid == NULL) {
+        char* signature = generate_method_signature(paramTypes, paramCount, returnType);
+        mid = (*env)->GetStaticMethodID(env, cls, methodName, signature);
+        if (mid == NULL) {
+            printf("Method not found\n");
```

Comment:
printf is ok for intermediate versions (but even then you possibly will go for the runtime_warning), for later you possibly want a COBOL exception to be raised